### PR TITLE
Move code collecting EnsoGL stats out of Sampler trait implementations

### PR DESCRIPTION
**Pull request by akavel**
### Pull Request Description

The goal of this PR is to decouple the code calculating EnsoGL runtime stats (notably, FPS, frame time, and WASM memory usage) from `Sampler` trait implementations (this trait is intended only to be responsible for post-processing stats data for display/view purposes).

Apart from the general improvement in separation of responsibilities, a longer-term goal for this change is to make it possible to access all the stats in any code that doesn't need the `Sampler` functionalities (expected to be needed for https://www.pivotaltracker.com/story/show/181093832 and https://www.pivotaltracker.com/story/show/181093601).

#### Visuals

The GUI that could be impacted by the changed code is the performance monitor panel, accessible with a `ctrl-option-tilde` keyboard shortcut in the IDE.

A screenshot of the Monitor panel **after** applying the PR, taken in IDE built with `./run dist`:

Screenshot 2022-02-02 at 12 38 09

A video recording of the panel taken in the same IDE (note that the process of recording itself has an impact on the FPS):

https://user-images.githubusercontent.com/273837/152157638-06d4e7fc-70ab-4f3e-be86-d4f7e2a327d8.mov

A screenshot of the Monitor panel **before** applying the PR, taken in IDE built with `./run dist`:

Screenshot 2022-02-02 at 12 25 24

A video recording of the panel taken in the same IDE:

https://user-images.githubusercontent.com/273837/152157916-11ac5287-5b37-4437-854c-9b458c666749.mov

https://www.pivotaltracker.com/story/show/181093700

### Important Notes

 - this change is split out from #3205

[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs:409]: You removed the function `begin` which was described as "Start measuring the data.". You moved the measurements out of Panel. You left the function `end` with the description "Stop measuring the data.". This is bad:

1. Panel is not measuring data anymore, so this is confusing for the reader.
2. The function `end` is not named correctly, because it doesn't provide the reader with enough semantic information – what does "end" mean if there was never "start"?

Please always change the naming and documentation of functions in structures that you change, so they reflect the changes.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: Why this structure is defined by hand instead of using macro, as almost all other structures?

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs:747]: After all these changes, this code is so similar to all other samplers, that instead of writing it by hand, we should imo add an option to the macro for conditional checks here and use macro to generate all the samplers instead. Please change the macro to allow passing a conditional value here instead.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs:784]: Why this change was made? Probably there is a reason for that, but I don't see it yet.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: This code is very repetitive. Please extract it to a function or a macro.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs:856]: these tests are top-notch, good job! ❤️

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: This not is not understandable by the reader. What does it mean that it "would throw off inner logic of Fps.". Also, it is strange that something bad will happen if the time is 0 – I'm not feeling comfortable with such a design, as I don't see any reason why JS timing could not return 0 in some cases.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: This needs to better explained. The sentence "FPS takes into account delays between frames" does not explain why this happens. Provide here a more detailed explanation, maybe with an example, please. A good example is that if for example the frame time was 1ms, we don't want to show crazy FPS, because the next frame will not be run immediately.

Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/monitor.rs]: I have another, pretty serious issue here – all of the samplers show the stats from the current frame, while the FPS stat shows the stats from the previous frame. This will be extremely confusing when debugging performance and this needs to be fixed. I would propose to make here a 2 step process – the monitor gets event (like function call) at the end of the frame, and gets a "draw" event on the beginning of the next frame. During the draw event, you have all information available, and all the stats can be updated in a synchronous manner.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs:155]: We are not using such checks nowhere in the code. Why it is used here?

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs:144]: this if block should have a short doc describing what it does. I know it was not in the previous codebase, but it should have one.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/monitor.rs]: @wdanilo Based on my understanding of the idea above, I would propose a slightly modified solution, where `StatsData` becomes available only as a result of `Stats::begin_frame(&self, time: f64) -> Option<StatsData>` (which would be returning previous frame's `StatsData` snapshot), evaluating to `None` on first call (because 1st frame does not have a predecessor). A `for` loop sampling and postprocessing stats data, and the `Monitor::draw()` call would then be moved from `Monitor::end()` to `Monitor::begin()`, such that the drawing would be done based on data loaded from the `StatsData` object. This would also imply that `Sampler` could and should lose a reference to `Stats`, instead being passed a `StatsData` object for sampling & postprocessing, and thus possibly become stateless. As a result, the Monitor Panel would be drawn at the beginning of a frame (instead of at the end) - following your proposal above - and would draw synchronized stats of the previous frame, except for 1st frame when it would need to skip the drawing operation, having received `None` from `begin_frame`.

The above proposal would also resolve a few other concerns expressed in this review (resolving the issue with "0.0" time being treated specially, as well as removing a need to add an extra conditional value to the `stats_sampler!` macro).

Given that the above proposal would necessitate refactoring in: `Monitor`, `Sampler` trait + `stats_sampler!` macro, `PanelData`, `Stats` and `StatsData`, I estimate the time this change would take at 1-2 days, taking into account also initial Q/A done as part of implementation, and possibly new unit tests. Based on that, I would like to ask for a recommendation how to proceed:

 - a) It can be noticed, that the issue we're discussing here is not a regression introduced by this PR, but it was present in the code before this PR. As such, it is a bug we need to solve regardless if we merge this PR or not. Taking this into account, we could split off a separate Bug task for that in Pivotal, in the Profiling EPIC, and remove those concerns from the current PR review, focusing in it just on refactoring current code, and prioritizing the Bug separately in the EPIC.
 - b) We could keep it as part of this PR, taking into consideration that together with remaining changes requested in this PR review, it will delay delivery of this PR and Task by est. 2 days.

What is your suggestion how to proceed with this? I consulted with @MichaelMauderer, who leans toward adding a new task, but also suggested I ask for your opinion?

Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/monitor.rs]: I feel that the solution a) would just make it way longer and way more painful than necessary. We are refactoring a code, we see that it does not work and it was pointed in the review. Merging it now, describing a new issue, starting working on it ... makes the whole process super formal, and would force us to spend so much additional time just on making "paperwork" that does not bring us to the final solution.

My question here is – what is the delivery time difference between points a) and b)? How much additional time we will waste when choosing point a)?

Regarding the proposed changes, I was not checking them deeply, but they seem reasonable on the first glance. The only thing I'd change is that I'd not return `Option<StatsData>`, just `StatsData` - for the first time it can be a default value. It makes sense - the first frame has 0 fps, 0 men usage, etc.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/stats.rs:155]: It's used here because `wasm_bindgen::memory()` panics in non-wasm_bindgen code, making it impossible to use `end_frame` in regular tests, as long as it contains WASM memory calculation. Alternatives I considered:

 1. Drop the check and make the tests be `#[wasm_bindgen_test]` instead of `#[test]` - however, this makes running the tests harder and slows down development iteration time.
 2. Use a compilation guard attribute on a block of code, instead of the `if cfg!(...)`:

        #[cfg(target = "wasm32")]
        {
            let memory: ...
            ...
        }

 3. Move the calculation of `wasm_memory_usage` outside of `end_frame()`, to a separate new method, and call this method externally from somewhere - though I didn't have a good idea what would be a good place to call it from.

Would you suggest I pick any other of the alternatives proposed above, or add a comment in code explaining the check, or do something else still?

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/monitor.rs:784]: We're introducing `f64` stats in this PR (`fps`, `frame_time`) - in those cases, `$stats_method()` evaluates to `f64`, and `f64 as f64` results in a warning about trivial numeric cast. I'm aware of the following options to work around this warning:

 - a) Keep the code as it was, but add `#![allow(trivial_numeric_casts)]` above, covering the whole expression, and explain in a comment why we use it;
 - b) Split out just the `self.stats.$stats_method() as f64` to a separate helper variable, and add `#![allow(trivial_numeric_casts)]` above it and add a comment with explanation;
 - c) Use `.as_()` provided by `num_traits::cast::AsPrimitive`, which works correctly (identity transformation) for `f64` without emitting a warning, which can be seen as an explicit statement of intent. Though, we could also add a comment explaining the reason why we're using it above, if you believe that's preferable.

Notably, we can't use `.into()` or `.try_into()` here, as they have different semantics for conversion of integer types (esp. `u64` and `usize`) into f64 vs. `as f64`. Specifically, `as f64` (and `.as_()`) perform an approximation if the converted type cannot fit into `f64` without loss of precision - this behavior sounds reasonable for use in stats - whereas `.try_into()` returns an error, forcing us to do approximation (or any other handling) by some other means.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/monitor.rs:747]: I plan to do this in https://www.pivotaltracker.com/story/show/181140499 - this sampler should then get simplified even further, making it possible to generate it with the macro without a need to add a conditional check.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/monitor.rs]: I worked on improving the comment; would you find the new phrasing good enough?

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/monitor.rs]: **Re. 1:** I attempted to improve the comment; would you find the new phrasing better?

**Re. 2:** As we discussed further, I created https://www.pivotaltracker.com/story/show/181140499 and plan to fix the calculation bug in `Fps` code (it was not throwing/panicking, but it was skipping some calculations - I now understand that the original phrasing I used in the comment here could be understood in a different way than I intended), which existed in old code in the same shape, through work on that Bug.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/monitor.rs:409]: I changed the function name to `sample_and_postprocess()`, using the meaning of "sample" as a verb (as in: "to sample"); however, I also considered `fetch_and_postprocess()` - would you maybe find that alternative (or some other one) easier to understand here?

Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/monitor.rs:409]: I like these names! Currently I cannot think of anything better.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/monitor.rs:747]: ok!

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/monitor.rs:784]: Perfect! I'm feeling right now bad that you spent so much time on explaining this to me. Thank you for doing that, the explanation is way more amazing that I was expecting it to be! The solution makes perfect sense and I also think it's the best out of the options!

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/stats.rs:155]: There is yet another alternative – @vitvakatu is right now working on allowing all wasp-related functions to return something (like 0) instead of panicking when compiled to a native code. We need exactly this thing in so many places, that handling it here differently than in other places doesn't make much sense. However, as Ilya's PR is not ready yet, let's leave it this way with an explanation and TODO pointing to Ilya's task explaining that we should move it to the new solution as soon as it will be available. How does it sound to you?

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs:853]: why this needs to be a macro and cannot be implemented just as a function?

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/stats.rs:144]: It's perfect

**Review by wdanilo — APPROVED** [review]
Amazing! It looks like it's mergeable ❤️

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/monitor.rs:853]: I started by making it a function (see: https://github.com/enso-org/enso/pull/3245/commits/58faf76bad15879ef10cc99a32281eba50cc41a6), but there was still some repetitiveness left due to `matches!` not being a function but a macro, and thus inability to pass just `$expected_check` as argument - the way I found around it was to pass a lambda, but it seemed rather clunky:

```rust
test.advance_frame(10.0, 6.0, |check| matches!(check, ValueCheck::Correct), 10.0);
```

Going with a macro doesn't seem to really be more complex, and the result seems easier on the eyes. I can bring back a function if you prefer?

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/stats.rs:155]: I love this idea; just added the TODO, thanks! ❤️

**Comment by akavel** [comment]
(*Process note:* the code-review is now accepted, therefore the PR is waiting for Review by @MichaelMauderer on Task https://www.pivotaltracker.com/story/show/181093700 before  merge.)

<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/181093700">Move code collecting EnsoGL stats out of Sampler trait implementations - Pivotal Tracker</a></strong></div></blockquote>

**Pull request merged**
